framework/doc/content/source/variables/ArrayMooseVariable.md

```diff
@@ -0,0 +1,89 @@
+# ArrayMooseVariable
+
+An array variable is define as a set of standard field variables with the same finite element family and order.
+Each standard variable of an array variable is referred to as a component of the array variable.
+An array kernel is a MOOSE kernel operating on an array variable and assembles the residuals and Jacobians for all the components of the array variable.
+A sample array kernel can be found at [ArrayDiffusion.md].
+The purpose of having array kernels is to reduce the number of kernels when the number of components of an array variable is large (potentially hundreds or thousands) and to avoid duplicated operations otherwise with lots of standard kernels.
+Array kernel can be useful for radiation transport where an extra independent direction variable can result into large number of standard variables.
+Similarly as array kernel, we have array initial conditions, array boundary conditions, array DG kernels, array interface kernels, array constraints, etc.
+
+The design:
+
+- to use variable groups in [libMesh] to group components in an array variable together.
+- to use template to avoid code duplication with standard and vector variables.
+- to use Eigen::Matrix to hold local dofs, solutions on quadrature points for an array variable to ease the local operations.
+- to use dense matrices or vectors as standard or vector variables with proper sizes for holding the local Jacobians and residuals that are to be assembled into a global Jacobian matrix and a global residual vector.
+
+The following map is useful for understanding the template:
+
+| OutputType          |OutputShape           |OutputData|
+|-|-|-|
+| Real                |Real                  |Real|
+| RealVectorValue     |RealVectorValue       |Real|
+| RealArrayValue      |Real                  |RealArrayValue|
+
+The three rows correspond to standard, vector and array variables.
+OutputType is the data type used for templating.
+RealArrayValue is a typedef in [MooseTypes.h] as *Eigen::Matrix<Real, Eigen::Dynamic, 1>*.
+OutputShape is for the type of shape functions and OutputData is the type of basis function expansion coefficients that are stored in the solution vector.
```

Comment:
One solution vector only holds one type (e.g. `Real`), so I'm a little unclear on what the last part of this sentence means. Can you explain a little more?

Reply:
Oh, this should be the `that are stored in the moose array variable, which are grabbed from the solution vector.`